### DISCUSSION TITLE
Is it possible to write a nameserver with Deno?

### DISCUSSION THREAD

**Discussion by NetOpWibby** [denoland/deno#14881] [Q&A] [answered]
I'm attempting to start a nameserver with my refactored version of [named-server](https://github.com/artur99/named-server) for Deno. However, all my errors are coming from [the standard library](https://deno.land/std@0.141.0/node/internal_binding/cares_wrap.ts).

Alas.

EDIT: I also tried just using the `named-server` library, same result.

```ts

import { default as named } from "https://esm.sh/named-server@1.0.3";

const server = named.createServer({ name: "yup" });

server.on("query", (query: any) => {
  const domain = query.name();

  if (domain === "registry.neuenet") {
    query.addAnswer(domain, new named.ARecord("143.110.239.106"), 604800);
    query.addAnswer(domain, new named.AAAARecord("2001:19f0:6001:412e:5400:03ff:fe7f:978d"), 604800);
  }

  server.send(query);
});

server.on("clientError", (error: Error) => {
  console.group("\n::: clientError");
  console.error(error.toString());
  console.groupEnd();
});

server.on("uncaughtException", (error: Error) => {
  console.group("\n::: uncaughtException");
  console.error(error.toString());
  console.groupEnd();
});

server.listen(9999, "127.0.0.1", () => console.log("DNS server started on port 9999"));
```

**Comment by NetOpWibby** [comment]
Truncated error messages:

```
...

TS2345 [ERROR]: Argument of type '"CAA"' is not assignable to parameter of type 'RecordType'.
    this.#query(name, "CAA").then(({ code, ret }) => {
                      ~~~~~
    at https://deno.land/std@0.143.0/node/internal_binding/cares_wrap.ts:374:23

TS2694 [ERROR]: Namespace 'Deno' has no exported member 'CAARecord'.
      const records = (ret as Deno.CAARecord[]).map(
                                   ~~~~~~~~~
    at https://deno.land/std@0.143.0/node/internal_binding/cares_wrap.ts:375:36

TS2345 [ERROR]: Argument of type '"NAPTR"' is not assignable to parameter of type 'RecordType'.
    this.#query(name, "NAPTR").then(({ code, ret }) => {
                      ~~~~~~~
    at https://deno.land/std@0.143.0/node/internal_binding/cares_wrap.ts:412:23

TS2694 [ERROR]: Namespace 'Deno' has no exported member 'NAPTRRecord'.
      const records = (ret as Deno.NAPTRRecord[]).map(
                                   ~~~~~~~~~~~
    at https://deno.land/std@0.143.0/node/internal_binding/cares_wrap.ts:413:36

TS2345 [ERROR]: Argument of type '"NS"' is not assignable to parameter of type 'RecordType'.
    this.#query(name, "NS").then(({ code, ret }) => {
                      ~~~~
    at https://deno.land/std@0.143.0/node/internal_binding/cares_wrap.ts:431:23

TS2345 [ERROR]: Argument of type '"SOA"' is not assignable to parameter of type 'RecordType'.
    this.#query(name, "SOA").then(({ code, ret }) => {
                      ~~~~~
    at https://deno.land/std@0.143.0/node/internal_binding/cares_wrap.ts:451:23

TS2694 [ERROR]: Namespace 'Deno' has no exported member 'SOARecord'.
          ret[0] as Deno.SOARecord;
                         ~~~~~~~~~
    at https://deno.land/std@0.143.0/node/internal_binding/cares_wrap.ts:456:26
```

**Reply by NetOpWibby** [reply]
_sigh_

**Reply by cjihrig** [reply]
What version of the CLI are you using?

**Reply by NetOpWibby** [reply]
1.21.2

**Reply by cjihrig** [reply]
It seems like something is importing a newer version of the standard library (0.143.0) than the one that shipped with your version of the CLI (0.138.0). You can upgrade to a newer version of the CLI (v1.22.3 or later based on the version of the standard library) or pin your standard library version to 0.138.0 or older.

**Reply by NetOpWibby** [reply]
I think that fixed my issues within Deno proper, thanks! I'm now dealing with another issue:

```sh
error: Uncaught TypeError: Cannot read properties of undefined (reading 'Resolver')
    at https://esm.sh/v78/cacheable-lookup@5.0.4/deno/cacheable-lookup.js:2:1082
    at https://esm.sh/v78/cacheable-lookup@5.0.4/deno/cacheable-lookup.js:2:681
    at https://esm.sh/v78/cacheable-lookup@5.0.4/deno/cacheable-lookup.js:2:5617
```

I think this has to do with modules I'm importing from `esm.sh` and not Deno though.

Thanks again @cjihrig!

**Comment by NetOpWibby** [comment, marked as answer]
> It seems like something is importing a newer version of the standard library (0.143.0) than the one that shipped with your version of the CLI (0.138.0). You can upgrade to a newer version of the CLI (v1.22.3 or later based on the version of the standard library) or pin your standard library version to 0.138.0 or older.

**Comment by NetOpWibby** [comment]
Hey @NetOpWibby, it's you, from the future.

You forked a project to work with Deno and modularized its codebase: https://deno.land/x/packet

You also have a custom nameserver using `packet` with a GraphQL API backed by EdgeDB, running in production. Huzzah! There's still some kinks to iron out but the answer to your original question is an emphatic YES.

**Reply by Corei13** [reply]
Hey @NetOpWibby, thank you for starting this thread and keep it going and finally resolve it yourself! Would you be open to help a fellow out on a similar journey to make their own nameserver in Deno by leaving some pointers?

**Reply by NetOpWibby** [reply]
@Corei13 The packet module should be enough to get started. The nameserver doesn't handle DNSSEC (which is where I got stuck and decided to go with PowerDNS to continue with my project). When I figure that out I'll open-source it. I'd rather not deal with feature requests and issues for it anytime soon.
